features/entire-workflow-versioning.md

```diff
@@ -0,0 +1,68 @@
+# Entire Workflow Versioning
+
+### Problem statement
+
+The current method of versioning Workflows is complex and error prone.
+By default, there's a good chance that upgrading a worker will result in workflows being stuck due to non-determinism errors.
+Recovering from this situation is not trivial because a user now has histories generated by mulitple versions.
+
+### Proposal
+
+Entire Workflow versioning is an approach that'll greatly simplify the current Workflow versioning situation.
+
+- We add a concept of a central repository for Workflow code
+- Users deploy their Workflows directly to the central repository
+- When a worker encounters the first Workflow Task in history it decides which Workflow code it needs to run
+  - New Workflows always use the "latest" version
+  - Workflows that were started with an old code version use that same version by default (unless there's an existing patch version (more on that below))
+
+### How does it work?
+
+- In JavaScript (and probably WASM in the future) we bundle the Workflow code with its dependencies into a single file
+- In Java we may use Jars (pending POC)
+- A user uploads their bundle to a central repository stating wheter it's the "latest" version or if it patches an older version
```

Comment:
This is very close to serverless/FaaS and I think we should explore the landscape here. Many of them have solved similar versioning issues with uploaded software.